velox/exec/Spiller.cpp

```diff
@@ -0,0 +1,402 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/exec/Spiller.h"
+
+#include "velox/common/base/AsyncSource.h"
+
+#include <folly/ScopeGuard.h>
+
+namespace facebook::velox::exec {
+
+void Spiller::extractSpill(folly::Range<char**> rows, RowVectorPtr* resultPtr) {
+  if (!*resultPtr) {
+    *resultPtr = std::static_pointer_cast<RowVector>(
+        BaseVector::create(rowType_, rows.size(), &spillPool()));
+  } else {
+    (*resultPtr)->prepareForReuse();
+    (*resultPtr)->resize(rows.size());
+  }
+  auto result = resultPtr->get();
+  auto& types = container_.columnTypes();
+  for (auto i = 0; i < types.size(); ++i) {
+    container_.extractColumn(rows.data(), rows.size(), i, result->childAt(i));
+  }
+  auto& aggregates = container_.aggregates();
+  auto numKeys = types.size();
+  for (auto i = 0; i < aggregates.size(); ++i) {
+    aggregates[i]->finalize(rows.data(), rows.size());
+    aggregates[i]->extractAccumulators(
+        rows.data(), rows.size(), &result->childAt(i + numKeys));
+  }
+}
+
+namespace {
+// A stream of ordered rows being read from the in memory
+// container. This is the part of a spillable range that is not yet
+// spilled when starting to produce output. This is only used for
+// sorted spills since for hash join spilling we just use the data in
+// the RowContainer as is.
+class RowContainerSpillStream : public SpillStream {
+ public:
+  RowContainerSpillStream(
+      RowTypePtr type,
+      int32_t numSortingKeys,
+      memory::MemoryPool& pool,
+      Spiller::SpillRows&& rows,
+      Spiller& spiller)
+      : SpillStream(std::move(type), numSortingKeys, pool),
+        rows_(std::move(rows)),
+        spiller_(spiller) {
+    if (!rows_.empty()) {
+      nextBatch();
+    }
+  }
+
+  uint64_t size() const override {
+    return 0;
+  }
+
+ private:
+  void nextBatch() override {
+    // Extracts up to 64 rows at a time. Small batch size because may
+    // have wide data and no gain in larger.when the caller will go
+    // over aggregations row by row.
+    static constexpr vector_size_t kMaxRows = 64;
+    constexpr uint64_t kMaxBytes = 4 << 20;
+    size_t bytes = 0;
+    vector_size_t numRows = 0;
+    auto limit = std::min<size_t>(rows_.size() - nextBatchIndex_, kMaxRows);
+    assert(!rows_.empty());
+    for (; numRows < limit; ++numRows) {
+      bytes += spiller_.container().rowSize(rows_[nextBatchIndex_ + numRows]);
+      if (bytes > kMaxBytes) {
+        ++numRows;
```

Comment:
nit: perhaps, comment to explain why we need to increment numRows. Is this to ensure that we spill at least one row?